src/main/java/com/google/firebase/auth/FirebaseAuth.java

```diff
@@ -342,28 +357,43 @@ public String createCustomToken(@NonNull String uid,
       final String uid, final Map<String, Object> developerClaims) {
     checkNotDestroyed();
     checkArgument(!Strings.isNullOrEmpty(uid), "uid must not be null or empty");
-    checkArgument(credentials instanceof ServiceAccountCredentials,
-        "Must initialize FirebaseApp with a service account credential to call "
-            + "createCustomToken()");
+    final FirebaseTokenFactory tokenFactory = ensureTokenFactory();
     return new CallableOperation<String, FirebaseAuthException>() {
       @Override
       public String execute() throws FirebaseAuthException {
-        final ServiceAccountCredentials serviceAccount = (ServiceAccountCredentials) credentials;
-        FirebaseTokenFactory tokenFactory = FirebaseTokenFactory.getInstance();
         try {
-          return tokenFactory.createSignedCustomAuthTokenForUser(
-              uid,
-              developerClaims,
-              serviceAccount.getClientEmail(),
-              serviceAccount.getPrivateKey());
-        } catch (GeneralSecurityException | IOException e) {
+          return tokenFactory.createSignedCustomAuthTokenForUser(uid, developerClaims);
+        } catch (IOException e) {
           throw new FirebaseAuthException(ERROR_CUSTOM_TOKEN,
               "Failed to generate a custom token", e);
         }
       }
     };
   }
 
+  private FirebaseTokenFactory ensureTokenFactory() {
+    FirebaseTokenFactory result = this.tokenFactory.get();
+    if (result == null) {
+      synchronized (lock) {
+        result = this.tokenFactory.get();
```

Comment:
If I remember my Java 101 class correctly, then you don need to use an AtomicReference here since you are grabbing a lock already.

http://www.cs.umd.edu/users/pugh/java/memoryModel/jsr-133-faq.html#synchronization
"Synchronization ensures that memory writes by a thread before or during a synchronized block are made visible in a predictable manner to other threads which synchronize on the same monitor. "

Reply:
Then we'd have to make the variable `volatile`, to get the double-checked locking semantics to work correctly: https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

I don't really have a preference. Drop the AtomicRef and switch to a volatile variable?